runtime/src/chain_ext.rs

```diff
@@ -1,167 +1,165 @@
 
-impl<T: Config> RegisteredChainExtension<T> for Extension<T>
-where
-	<T as SysConfig>::AccountId: AsRef<[u8; 32]>,
-{
-	const ID: u16 = 1;
+                let message: VersionedXcm<CallOf<T>> = env.read_as_unbounded(env.in_len())?;
+                let message = Box::new(message);
+                let origin = frame_system::Origin::<T>::Signed(env.ext().address().clone());
+                let max_weight = env.ext().gas_meter().gas_left();
+
+                let charged = env.charge_weight(<T as pallet_xcm::Config>::WeightInfo::execute())?;
+                let res = pallet_xcm::Pallet::<T>::execute(origin.into(), message, max_weight);
+                if let Some(weight) = res.map_or_else(|res| res.post_info, identity).actual_weight {
+                    env.adjust_weight(charged, weight);
+                }
+
+                res.map_err(|err| err.error)?;
+            }
+
+            // Send an XCM message from the contract to the specified destination.
+            Command::Send => {
+                log::debug!(target: LOG_TARGET, "Send XCM message");
+                let mut env = env.buf_in_buf_out();
+                let origin = frame_system::Origin::<T>::Signed(env.ext().address().clone());
+                let SendInput { dest, msg } = env.read_as_unbounded(env.in_len())?;
+                env.charge_weight(<T as pallet_xcm::Config>::WeightInfo::send())?;
+                pallet_xcm::Pallet::<T>::send(origin.into(), Box::new(dest), Box::new(msg))?;
+            }
+
+            // Create a new query, using the contract's address as the responder.
+            Command::NewQuery => {
+                log::debug!(target: LOG_TARGET, "Create new XCM query");
+                let mut env = env.buf_in_buf_out();
+                let len = env.in_len();
+                let NewQueryInput {
+                    timeout,
+                    match_querier,
+                }: NewQueryInput = env.read_as_unbounded(len)?;
+
+                let responder = MultiLocation {
+                    parents: 0,
+                    interior: Junctions::X1(Junction::AccountId32 {
+                        network: None,
+                        id: *env.ext().address().as_ref(),
+                    }),
+                };
+
+                // TODO Charge weight
+
+                let query_id = <pallet_xcm::Pallet<T> as QueryHandler>::new_query(
+                    responder,
+                    timeout.into(),
+                    MultiLocation::try_from(match_querier)
+                        .map_err(|_| XCMError::<T>::BadVersion)?,
+                );
+
+                log::debug!(target: LOG_TARGET, "new query id {query_id}");
+                query_id.using_encoded(|q| env.write(q, true, None))?;
+            }
+
+            // Attempt to take a response for the specified query.
+            Command::TakeResponse => {
+                log::debug!(target: LOG_TARGET, "Take XCM query response");
```

Comment:
a better option here is maybe to make the type Encode, Decode upstream and just return that to the caller...